llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp

```diff
@@ -1331,11 +1331,17 @@ bool LoopVectorizationLegality::canVectorizeWithIfConvert() {
     // we restrict this to loads; stores are more complicated due to
     // concurrency restrictions.
     ScalarEvolution &SE = *PSE.getSE();
+    SmallVector<const SCEVPredicate *, 4> Predicates;
     for (Instruction &I : *BB) {
       LoadInst *LI = dyn_cast<LoadInst>(&I);
+      // Pass the Predicates pointer to isDereferenceableAndAlignedInLoop so
+      // that it will consider loops that need guarding by SCEV checks. The
+      // vectoriser will generate these checks if we decide to vectorise.
       if (LI && !LI->getType()->isVectorTy() && !mustSuppressSpeculation(*LI) &&
-          isDereferenceableAndAlignedInLoop(LI, TheLoop, SE, *DT, AC))
+          isDereferenceableAndAlignedInLoop(LI, TheLoop, SE, *DT, AC,
+                                            &Predicates))
         SafePointers.insert(LI->getPointerOperand());
+      Predicates.clear();
```

Comment:
I might not understand the full underlying magic, but I need help seeing a generation of the required predicates. As far as I understand, it is legal to put a pointer into `SafePointer` if `Predicates` would be generated before the vectorised loop, but I don't see any other code that collects those predicates again or generates them.

Reply:
I realise it looks a bit confusing, but when we call `isDereferenceableAndAlignedInLoop` all that we really care about is whether or not we can dereference these loads unconditionally if we assume the predicates are set up correctly. In fact, if we actually decide to vectorise the loop we will collect *all* the predicates required for all blocks in the loop. This happens for free because at some point we call `PSE.getBackedgeTakenCount()` and the PredicatedScalarEvolution class will keep track of all the predicates, which will get code-generated later. I added the `Predicates` argument to `isDereferenceableAndAlignedInLoop` in case someone actually wants to inspect each predicate, but in this particular case it's safe to drop them.

Reply:
Thanks for the explanation, I thought that something like this was happening, but I couldn't pin it to the exact responsible code